web_external/js/views/widgets/AddSourceWidget.js

```diff
@@ -6,39 +6,63 @@ minerva.views.AddSourceWidget = minerva.View.extend({
         'submit #m-add-source-form': function (e) {
             e.preventDefault();
 
+            // TODO better to have single buttons instead of radio?
+            // That way user only has to click once instead of radio + Add
             var sourceType = $('#m-add-source-form input:radio:checked').attr('id');
+            // get sourcetype from m-sourcetype-source
+            sourceType = sourceType.split('-')[1];
             var container = $('#g-dialog-container');
 
-            if (sourceType === 'm-wms-source') {
-                this.wmsSourceWidget = new minerva.views.AddWmsSourceWidget({
-                    el: container,
-                    title: 'Enter WMS Source details',
-                    noParent: true,
-                    collection: this.collection,
-                    parentView: this.parentView
-                }).render();
-            } else if (sourceType === 'm-elasticsearch-source') {
-                this.elasticsearchSourceWidget = new minerva.views.AddElasticsearchSourceWidget({
-                    el: container,
-                    title: 'Enter Elasticsearch Source details',
-                    noParent: true,
-                    collection: this.collection,
-                    parentView: this.parentView
-                }).render();
+            var settings = {
+                el: container,
+                noParent: true,
+                collection: this.collection,
+                parentView: this.parentView
+            };
+
+            if (_.has(this.sourceTypes, sourceType)) {
+                var widget = new this.sourceTypes[sourceType].widget(settings);
+                widget.render();
             } else {
                 console.error('Unknown source type');
             }
+        },
+        'click .m-add-source-radio': function (e) {
+            $(e.currentTarget).find('input:radio').first().prop('checked', true);
         }
     },
 
     initialize: function (settings) {
         this.collection = settings.collection;
+        // TODO would be nice if new source types could register themselves,
+        // perhaps with a method on the minerva object, that we could then
+        // query here.  All the source types register themselves upon definition,
+        // and we query here upon instantation.  This registration process
+        // should also add them to the SourceCollection model function.
+        this.sourceTypes = {
+            wms: {
+                label: 'WMS',
+                icon: 'icon-layers',
+                widget: minerva.views.AddWmsSourceWidget
+            },
+            elasticsearch: {
+                label: 'Elasticsearch',
+                icon: 'icon-search',
+                widget: minerva.views.AddElasticsearchSourceWidget
+            },
+            s3: {
+                label: 'S3 bucket',
+                icon: 'icon-cloud',
+                widget: minerva.views.AddS3SourceWidget
+            }
+        };
     },
```

Comment:
Since these are going to be in the SourceCollection and they're hardcoded, does it make sense to put them there now? I understand this is related to their UI, but it seems pretty counter-intuitive that the types of sources are registered in a widget view.

This way we could add the source model attribute to this definition and remove the if/else statements in SourceCollection.

This might not make sense since we're planning to refactor a lot of this soon anyway, in which case - disregard.